# Dublin House Prices by Post Code
## Part 2 - Cleaning the Data
In [Part 1](), we took a download of house price data for Dublin in 2016 from the [Property Price Register](https://www.propertypriceregister.ie/website/npsra/pprweb.nsf/page/ppr-home-en) website, and used a `Google Maps Client` to add some further detail to our data.

Now we're going to examine that data, and see if we can further fill our data set.

We pickled our data as two `pandas` data frames, one with good address (meaning, addresses for which our `Google Maps Client` gave us all the detail we wanted, and a second with those address for which our `Google Maps Client` was less successful. We'll upload them now, and look at them.

In [1]:
import pickle
import pandas as pd

with open('../library/pickle/2016/houses_dublin_2016_pickle', 'r') as f:
    good_addresses = pickle.load(f)
    
with open('../library/pickle/2016/bad_add_dublin_2016_pickle', 'r') as f:
    bad_addresses = pickle.load(f)

In [2]:
good_addresses.describe()

Lat           Lon
count  10514.000000  10514.000000
mean      53.352123     -6.263577
std        0.071453      0.089437
min       53.202946     -6.501789
25%       53.297334     -6.317503
50%       53.343143     -6.255972
75%       53.385494     -6.199972
max       53.626983     -6.052932

In [3]:
bad_addresses.describe()

Date                                            Address  \
count         2244                                               2244   
unique         241                                               2227   
top     17/02/2016  APT.7 - DOREEN HOUSE, 317-319 BLACKHORSE AVENU...   
freq           208                                                  2   

       PostCode  County        Price FullMarketPrice   VAT  \
count      2244    2244         2244            2244  2244   
unique       24       2          922               3     3   
top              Dublin  €167,117.00              No    No   
freq        756    2243           56            2180  1516   

                                  Description  Size      gCheck  
count                                    2244  2244        2244  
unique                                      3     6           4  
top     Second-Hand Dwelling house /Apartment        NoResponse  
freq                                     1510  1510        1922

In [4]:
good_addresses.dtypes

Date                object
Address             object
PostCode            object
County              object
Price               object
FullMarketPrice     object
VAT                 object
Description         object
Size                object
Lat                float64
Lon                float64
gCheck              object
dtype: object

In [5]:
bad_addresses.dtypes

Date               object
Address            object
PostCode           object
County             object
Price              object
FullMarketPrice    object
VAT                object
Description        object
Size               object
gCheck             object
dtype: object

In [6]:
good_addresses.head()

Date                                            Address  PostCode  \
0  01/01/2016  34 Mountpleasant Terrace, Dublin 6, D06 YC58, ...  D06 YC58   
1  04/01/2016                       Saggart, Co. Dublin, Ireland             
2  04/01/2016                       Saggart, Co. Dublin, Ireland             
3  04/01/2016  2 Brighton Rd, Brighton Hall, Kerrymount, Dubl...  D18 A0C6   
4  04/01/2016  Moyola Mews, Churchtown Rd Lower, Churchtown L...             

   County          Price FullMarketPrice  VAT  \
0  Dublin    €170,000.00              No   No   
1  Dublin    €230,088.00              No  Yes   
2  Dublin    €321,586.00              No  Yes   
3  Dublin  €1,150,000.00              No   No   
4  Dublin    €330,000.00              No   No   

                             Description  \
0  Second-Hand Dwelling house /Apartment   
1          New Dwelling house /Apartment   
2          New Dwelling house /Apartment   
3  Second-Hand Dwelling house /Apartment   
4  Second-Hand Dwelling house /Apartment   

                                                Size        Lat       Lon  \
0                                                     53.328587 -6.261495   
1  greater than or equal to 38 sq metres and less...  53.280840 -6.443184   
2  greater than or equal to 38 sq metres and less...  53.280840 -6.443184   
3                                                     53.258165 -6.174641   
4                                                     53.298311 -6.253560   

  gCheck  
0   Good  
1   Good  
2   Good  
3   Good  
4   Good

In [7]:
bad_addresses.head()

Date                                          Address  \
0  Date of Sale (dd/mm/yyyy)                                          Address   
1                 04/01/2016        307 HAMPTON LODGE, BLOCK B, GRACE PARK RD   
2                 05/01/2016                 21 AMBER, THE GRANGE, STILLORGAN   
3                 05/01/2016         29 GREENVIEW, SEABROOK MANOR, STATION RD   
4                 05/01/2016  5 WILLOW GROVE, GREENPARK, CLONDALKIN DUBLIN 22   

      PostCode  County        Price        FullMarketPrice            VAT  \
0  Postal Code  County    Price (€)  Not Full Market Price  VAT Exclusive   
1     Dublin 9  Dublin  €315,000.00                     No             No   
2               Dublin  €280,000.00                     No             No   
3    Dublin 13  Dublin   €94,042.00                     No             No   
4    Dublin 22  Dublin  €265,000.00                     No             No   

                             Description                       Size  \
0                Description of Property  Property Size Description   
1  Second-Hand Dwelling house /Apartment                              
2  Second-Hand Dwelling house /Apartment                              
3  Second-Hand Dwelling house /Apartment                              
4  Second-Hand Dwelling house /Apartment                              

       gCheck  
0  NoResponse  
1      BadLat  
2  NoResponse  
3  NoResponse  
4  NoResponse

#### Bad v Good Addresses
The difference between bad and good addresss is the presence of coordinates in good addresses, and the absence of coordinates in the bad data. Otherwise, the data is quite the same. Therefore, although the `coordinateCollector` program divides the two data sets, it makes sense for us to unite them here into a single data frame, `houses`. We'll add two "empty" columns to `bad_addresses` for the missing coordinates to make the join nice and even.

You'll also notice that the first line of `bad_addresses` is a repeated header row. That will have to go, of course.

In [8]:
bad_addresses.drop(0, inplace=True) # Eliminate the repeated header row.

import numpy as np
bad_addresses['Lat'] = np.nan
bad_addresses['Lon'] = np.nan
houses = pd.concat([good_addresses, bad_addresses], ignore_index=True)

In [9]:
houses.head()

Address  County        Date  \
0  34 Mountpleasant Terrace, Dublin 6, D06 YC58, ...  Dublin  01/01/2016   
1                       Saggart, Co. Dublin, Ireland  Dublin  04/01/2016   
2                       Saggart, Co. Dublin, Ireland  Dublin  04/01/2016   
3  2 Brighton Rd, Brighton Hall, Kerrymount, Dubl...  Dublin  04/01/2016   
4  Moyola Mews, Churchtown Rd Lower, Churchtown L...  Dublin  04/01/2016   

                             Description FullMarketPrice        Lat       Lon  \
0  Second-Hand Dwelling house /Apartment              No  53.328587 -6.261495   
1          New Dwelling house /Apartment              No  53.280840 -6.443184   
2          New Dwelling house /Apartment              No  53.280840 -6.443184   
3  Second-Hand Dwelling house /Apartment              No  53.258165 -6.174641   
4  Second-Hand Dwelling house /Apartment              No  53.298311 -6.253560   

   PostCode          Price                                               Size  \
0  D06 YC58    €170,000.00                                                      
1              €230,088.00  greater than or equal to 38 sq metres and less...   
2              €321,586.00  greater than or equal to 38 sq metres and less...   
3  D18 A0C6  €1,150,000.00                                                      
4              €330,000.00                                                      

   VAT gCheck  
0   No   Good  
1  Yes   Good  
2  Yes   Good  
3   No   Good  
4   No   Good

In [10]:
houses.tail()

Address  County        Date  \
12752         APT 2, GLENANNE, KIMMAGE CROSS RD DUBLIN 6  Dublin  23/11/2016   
12753  110 CARRIGWOOD, BALLYCULLEN DR, FIRHOUSE DUBLI...  Dublin  24/11/2016   
12754  No.84, Roseland Avenue, Cualanor Upper Glenage...  Dublin  24/11/2016   
12755                  10 CADDELL, THE LINKS, STATION RD  Dublin  25/11/2016   
12756  No 10 Royal Canal Avenue, Royal Canal Park, Ra...  Dublin  25/11/2016   

                                 Description FullMarketPrice  Lat  Lon  \
12752  Second-Hand Dwelling house /Apartment              No  NaN  NaN   
12753  Second-Hand Dwelling house /Apartment              No  NaN  NaN   
12754          New Dwelling house /Apartment              No  NaN  NaN   
12755  Second-Hand Dwelling house /Apartment              No  NaN  NaN   
12756          New Dwelling house /Apartment              No  NaN  NaN   

        PostCode        Price  \
12752             €235,500.00   
12753  Dublin 24  €340,000.00   
12754             €651,982.00   
12755  Dublin 13  €270,000.00   
12756  Dublin 15  €325,991.19   

                                                    Size  VAT      gCheck  
12752                                                      No  NoResponse  
12753                                                      No  NoResponse  
12754             greater than or equal to 125 sq metres  Yes  NoResponse  
12755                                                      No  NoResponse  
12756  greater than or equal to 38 sq metres and less...  Yes  NoResponse

In [11]:
houses.dtypes

Address             object
County              object
Date                object
Description         object
FullMarketPrice     object
Lat                float64
Lon                float64
PostCode            object
Price               object
Size                object
VAT                 object
gCheck              object
dtype: object

#### Tidying Some Formats
We need to convert the date and price columns to datetime and float data types, respectively. The date conversion is straight-forward. The price less so, probably due to the _latin-1_ encoding - casting `astype(float)` doesn't work. But it still can be done, of course, even if we have to go the long way around.

In [12]:
houses['Date'] = pd.to_datetime(houses['Date'], dayfirst = True)

def price_destringer(someString):
    someOtherString = ''
    for s in someString:
        if s.isdigit() or s == '.':
            someOtherString = someOtherString + s
        else:
            pass
        
    return float(someOtherString)

houses['aPrice'] = houses['Price'].apply(price_destringer)
del(houses['Price'])
houses.rename(columns = {'aPrice':'Price'}, inplace = True)
houses.dtypes

Address                    object
County                     object
Date               datetime64[ns]
Description                object
FullMarketPrice            object
Lat                       float64
Lon                       float64
PostCode                   object
Size                       object
VAT                        object
gCheck                     object
Price                     float64
dtype: object

### The Postcodes
The _PostCode_ column contains a mix of data formats - the modern XXX XXXX format; the traditional Dublin 4-style format, and blanks. We're going to use a function, `tidy_postcodes()`, to create a second postcode column, PC, that will;
1. Return the first three characters of the new post code format if the extant post code is that format;
2. Return the first three characters of the new post code format if the extant post code is of the old-style Dublin post-code format, using the easy-to_create `short_postcode_dictionary`; and
3. Return glugger, from the Irish _ubh ghliogair_, a rotten egg, if the post code is neither of the above.

The first three characters of the new post codes are called the **Eircode Routing Keys** - I am indebted to Green Party Councillor for Dún Laoghaire, Ossian Smyth, for his excellent breakdown of these Eircode Routing keys on his website [here](http://www.ossiansmyth.ie/eircode-routing-keys/).

In [13]:
import re

short_postcode_mapper = {}
for x in range(1,25):
    myKey = 'Dublin {}'.format(x)
    myValue = 'D{:02}'.format(x)
    short_postcode_mapper[myKey] = myValue
short_postcode_mapper['Dublin 6W'] = 'D06'
short_postcode_mapper['Dublin 6w'] = 'D06'

new_postcode = re.compile('[A-Z][0-9]{2}')
old_postcode = re.compile('Dublin [0-9]{1,2}')

def tidy_postcode(somePostCode):
    if re.match(old_postcode, somePostCode):
        return short_postcode_mapper[somePostCode]
    elif re.match(new_postcode, somePostCode):
        return somePostCode[:3]
    else:
        return 'glugger'
    
houses['PC'] = houses['PostCode'].apply(tidy_postcode)
houses.PC.value_counts()

glugger    4822
D15         790
D04         532
D24         430
D07         428
D09         415
D12         397
D18         370
D08         322
D11         320
D16         317
D13         312
D06         306
D14         297
D03         289
D05         284
A94         273
A96         255
K78         250
K67         233
D22         208
D01         172
K32         164
K36         120
D10          95
D02          90
D20          78
D17          68
K34          47
K45          36
K56          17
A98          15
A45           2
A41           1
A86           1
A42           1
Name: PC, dtype: int64

And here we have a problem - we have 4,822 addresses for which we don't have an eirKey - a frustratingly higher number than the old-style post codes that existed in the *Property Price Register* data. But not to worry - we can use Machine Learning algorithms track down some more eirKeys in [Part 3]().

In [15]:
with open('../library/pickle/2016/houses_cleaned_after_google_2016_pickle', 'w') as f:
    pickle.dump(houses, f)